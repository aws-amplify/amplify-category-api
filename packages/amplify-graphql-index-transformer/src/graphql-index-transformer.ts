import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InvalidDirectiveError,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { IndexDirective } from '@aws-amplify/graphql-directives';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isListType, isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { generateKeyAndQueryNameForConfig, validateNotSelfReferencing } from './utils';

/**
 *
 */
export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];

  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', IndexDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as IndexDirectiveConfiguration,
      generateGetArgumentsInput(context.transformParameters),
    );

    /**
     * Impute Optional Fields
     * Start with sort key fields since both the name and queryField will work sort key fields into the generated name if necessary.
     */
    args.sortKeyFields = getOrGenerateDefaultSortKeyFields(args);
    args.name = getOrGenerateDefaultName(args);
    args.queryField = getOrGenerateDefaultQueryField(context, args);
    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

/**
 * Return the name if provided in our args, else
 * compute the name based on the field name, and sortKeyFields.
 */
const getOrGenerateDefaultName = (config: IndexDirectiveConfiguration): string => {
  const indexNameRegex = /^[A-Za-z0-9_\-\.]{3,255}$/;
  if (config.name) {
    if (!indexNameRegex.test(config.name)) {
      throw new Error(
        `The indexName is invalid. It should be between 3 and 255 characters. Only A–Z, a–z, 0–9, underscore characters, hyphens, and periods allowed.`,
      );
    }
    return config.name;
  }

  if (config.name === null) {
    throw new Error('Explicit null value not allowed for name field on @index');
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * Return the queryField if provided in our args, else
 * compute the queryField based on the field name, and sortKeyFields if the feature flag is enabled.
 */
const getOrGenerateDefaultQueryField = (
  context: TransformerSchemaVisitStepContextProvider,
  config: IndexDirectiveConfiguration,
): string | null => {
  const autoIndexQueryNamesIsEnabled = context.transformParameters.enableAutoIndexQueryNames;
  // Any explicit null will take effect, if enableAutoIndexQueryNames and no queryField is provide set to null for consistency
  if (config.queryField === null || (!autoIndexQueryNamesIsEnabled && !config.queryField)) {
    return null;
  }

  if (config.queryField) {
    return config.queryField;
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * sortKeyFields are optional so default to empty list,
 * if we get a raw object just wrap in an array,
 * else return the list which was provided correctly.
 */
const getOrGenerateDefaultSortKeyFields = (config: IndexDirectiveConfiguration): string[] => {
  if (!config.sortKeyFields) {
    return [];
  }
  if (!Array.isArray(config.sortKeyFields)) {
    return [config.sortKeyFields];
  }
  return config.sortKeyFields;
};

const validate = (config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void => {
  const { name, object, field, sortKeyFields } = config;
  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find((directive) => directive.name.value === 'model');

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${IndexDirective.name} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        const hasSortFields = peerDirective.arguments!.some(
          (arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0,
        );
        config.primaryKeyField = objectField;

        if (!hasSortFields && objectField.name.value === field.name.value) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === IndexDirective.name &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${IndexDirective.name} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }

    for (const peerDirective of objectField.directives!) {
      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0);

      if (
        !ctx.transformParameters.secondaryKeyAsGSI &&
        !hasSortFields &&
        objectField == config.primaryKeyField &&
        objectField.name.value === field.name.value
      ) {
        throw new InvalidDirectiveError(
          `Invalid @index '${name}'. You may not create an index where the partition key ` +
            'is the same as that of the primary key unless the index has a sort field. ' +
            'You cannot have a local secondary index without a sort key in the index.',
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums) || isListType(field.type)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums) || isListType(sortField.type)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
};
