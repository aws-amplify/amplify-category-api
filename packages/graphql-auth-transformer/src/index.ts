export * from './ModelAuthTransformer';
// No-op change to trigger re-publish
