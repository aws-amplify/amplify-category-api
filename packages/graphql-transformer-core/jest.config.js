const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
