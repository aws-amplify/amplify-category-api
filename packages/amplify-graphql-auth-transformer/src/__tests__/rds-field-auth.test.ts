import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { constructDataSourceStrategies, validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { mockSqlDataSourceStrategy, testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { parse } from 'graphql';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';
import { PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { BelongsToTransformer, HasOneTransformer } from '@aws-amplify/graphql-relational-transformer';
import { AuthTransformer } from '../graphql-auth-transformer';

describe('Verify RDS Model level Auth rules on queries:', () => {
  const mysqlStrategy = mockSqlDataSourceStrategy();

  it('should successfully transform different field auth rules', async () => {
    const validSchema = `
        type Post @model @auth(rules: [{ allow: private }]) {
            id: ID! @primaryKey @auth(rules: [{ allow: private }, { allow: public }])
            owner: String
            authors: [String]
            privateContent: String @auth(rules: [{ allow: private }])
            publicContent: String @auth(rules: [{ allow: public }])
            ownerContent: String @auth(rules: [{ allow: owner, operations: [create, read], identityClaim: "user_id" }])
            ownersContent: String @auth(rules: [{ allow: owner, ownerField: "authors", operations: [update, read] }])
        }
    `;

    const authConfig: AppSyncAuthConfiguration = {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
        userPoolConfig: {
          userPoolId: 'TEST_USER_POOL_ID',
        },
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'API_KEY',
        },
      ],
    };

    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new AuthTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, mysqlStrategy),
      authConfig,
      synthParameters: {
        identityPoolId: 'TEST_IDENTITY_POOL_ID',
      },
    });
    expect(out).toBeDefined();

    validateModelSchema(parse(out.schema));
    parse(out.schema);

    // Verify the field auth resolver slots
    expect(out.resolvers['Post.authors.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.authors.res.vtl']).toMatchSnapshot();

    expect(out.resolvers['Post.owner.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.owner.res.vtl']).toMatchSnapshot();

    expect(out.resolvers['Post.ownerContent.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.ownerContent.res.vtl']).toMatchSnapshot();

    expect(out.resolvers['Post.ownersContent.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.ownersContent.res.vtl']).toMatchSnapshot();

    expect(out.resolvers['Post.privateContent.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.privateContent.res.vtl']).toMatchSnapshot();

    expect(out.resolvers['Post.publicContent.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Post.publicContent.res.vtl']).toMatchSnapshot();
  });

  it('related models with same groups auth field name but different type should transform correctly', async () => {
    const validSchema = `
      type User 
        @model 
        @auth(rules: [
          { allow: groups, groups: ["Admin"] },
          { allow: public, operations: [get] },
          { allow: groups, groups: ["Dev"], operations: [read] }
          { allow: groups, groupsField: "groupsField", operations: [update, delete] }
        ])
      {
        id: String! @primaryKey
        name: String
        groupsField: String
        profile: Profile @hasOne(references: ["userId"])
      }
      type Profile 
        @model 
        @auth(rules: [
          { allow: groups, groups: ["Admin"] },
          { allow: public, operations: [list] },
          { allow: groups, groups: ["Dev"], operations: [get, create, update, delete] }, 
          { allow: groups, groupsField: "groupsField", operations: [read, create] }
        ])
      {
        id: String! @primaryKey
        details: String
        groupsField: [String]
        userId: String!
        user: User @belongsTo(references: ["userId"])
      }
    `;

    const authConfig: AppSyncAuthConfiguration = {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
        userPoolConfig: {
          userPoolId: 'TEST_USER_POOL_ID',
        },
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'API_KEY',
        },
      ],
    };

    const out = testTransform({
      schema: validSchema,
      transformers: [
        new ModelTransformer(),
        new AuthTransformer(),
        new PrimaryKeyTransformer(),
        new HasOneTransformer(),
        new BelongsToTransformer(),
      ],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, mysqlStrategy),
      authConfig,
      synthParameters: {
        identityPoolId: 'TEST_IDENTITY_POOL_ID',
      },
    });
    expect(out).toBeDefined();

    validateModelSchema(parse(out.schema));
    parse(out.schema);

    // groups field must be of type `string` in the generated `User` resolvers
    expect(out.resolvers['Mutation.updateUser.auth.1.res.vtl']).toMatchSnapshot();
    expect(out.resolvers['Mutation.updateUser.auth.1.res.vtl']).toEqual(expect.stringContaining('"groupsFieldType": "string"'));

    expect(out.resolvers['Mutation.deleteUser.auth.1.res.vtl']).toMatchSnapshot();
    expect(out.resolvers['Mutation.deleteUser.auth.1.res.vtl']).toEqual(expect.stringContaining('"groupsFieldType": "string"'));

    // groups field must be of type `string[]` in the generated `Profile` resolvers
    expect(out.resolvers['Mutation.createProfile.auth.1.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Mutation.createProfile.auth.1.req.vtl']).toEqual(expect.stringContaining('"groupsFieldType": "string[]"'));

    expect(out.resolvers['Query.listProfiles.auth.1.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Query.listProfiles.auth.1.req.vtl']).toEqual(expect.stringContaining('"groupsFieldType": "string[]"'));

    expect(out.resolvers['Query.getProfile.auth.1.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Query.getProfile.auth.1.req.vtl']).toEqual(expect.stringContaining('"groupsFieldType": "string[]"'));
  });
});
