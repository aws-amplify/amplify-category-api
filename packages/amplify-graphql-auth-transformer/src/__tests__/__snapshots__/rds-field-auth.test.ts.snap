// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify RDS Model level Auth rules on queries: related models with same groups auth field name but different type should transform correctly 1`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Admin\\"],
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"name\\", \\"groupsField\\", \\"profile\\"]
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"groupsFieldName\\": \\"groupsField\\",
  \\"groupsFieldType\\": \\"string\\",
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"name\\", \\"groupsField\\", \\"profile\\"]
}))
#set( $authResult = $util.authRules.mutationAuth($authRules, \\"update\\", $ctx.args.input, $ctx.result) )
#if( !$authResult || ($authResult && !$authResult.authorized) )
  $util.unauthorized()
#end
#if( $authResult && !$util.isNullOrEmpty($authResult.authorizedInput) )
  #set( $ctx.args.input = $authResult.authorizedInput )
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: related models with same groups auth field name but different type should transform correctly 2`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Admin\\"],
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"name\\", \\"groupsField\\", \\"profile\\"]
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"groupsFieldName\\": \\"groupsField\\",
  \\"groupsFieldType\\": \\"string\\",
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"name\\", \\"groupsField\\", \\"profile\\"]
}))
#set( $authResult = $util.authRules.mutationAuth($authRules, \\"delete\\", $ctx.args.input, $ctx.result) )
#if( !$authResult || ($authResult && !$authResult.authorized) )
  $util.unauthorized()
#end
#if( $authResult && !$util.isNullOrEmpty($authResult.authorizedInput) )
  #set( $ctx.args.input = $authResult.authorizedInput )
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: related models with same groups auth field name but different type should transform correctly 3`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Admin\\"],
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"details\\", \\"groupsField\\", \\"userId\\", \\"user\\"]
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Dev\\"],
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"details\\", \\"groupsField\\", \\"userId\\", \\"user\\"]
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"groupsFieldName\\": \\"groupsField\\",
  \\"groupsFieldType\\": \\"string[]\\",
  \\"groupClaim\\": \\"cognito:groups\\",
  \\"allowedFields\\":   [\\"id\\", \\"details\\", \\"groupsField\\", \\"userId\\", \\"user\\"]
}))
#set( $authResult = $util.authRules.mutationAuth($authRules, \\"create\\", $ctx.args.input, null) )
#if( !$authResult || ($authResult && !$authResult.authorized) )
  $util.unauthorized()
#end
#if( $authResult && !$util.isNullOrEmpty($authResult.authorizedInput) )
  #set( $ctx.args.input = $authResult.authorizedInput )
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: related models with same groups auth field name but different type should transform correctly 4`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Admin\\"],
  \\"groupClaim\\": \\"cognito:groups\\"
}))
$util.qr($authRules.add({
  \\"type\\": \\"public\\",
  \\"provider\\": \\"apiKey\\"
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"groupsFieldName\\": \\"groupsField\\",
  \\"groupsFieldType\\": \\"string[]\\",
  \\"groupClaim\\": \\"cognito:groups\\"
}))
#set( $authResult = $util.authRules.queryAuth($authRules) )
#if( !$authResult || ($authResult && !$authResult.authorized) )
  $util.unauthorized()
#end
#if( $authResult && !$util.isNullOrEmpty($authResult.authFilter) )
  #set( $ctx.stash.authFilter = $authResult.authFilter )
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: related models with same groups auth field name but different type should transform correctly 5`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Admin\\"],
  \\"groupClaim\\": \\"cognito:groups\\"
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"allowedGroups\\":   [\\"Dev\\"],
  \\"groupClaim\\": \\"cognito:groups\\"
}))
$util.qr($authRules.add({
  \\"type\\": \\"groups\\",
  \\"provider\\": \\"userPools\\",
  \\"groupsFieldName\\": \\"groupsField\\",
  \\"groupsFieldType\\": \\"string[]\\",
  \\"groupClaim\\": \\"cognito:groups\\"
}))
#set( $authResult = $util.authRules.queryAuth($authRules) )
#if( !$authResult || ($authResult && !$authResult.authorized) )
  $util.unauthorized()
#end
#if( $authResult && !$util.isNullOrEmpty($authResult.authFilter) )
  #set( $ctx.stash.authFilter = $authResult.authFilter )
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 1`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"private\\",
  \\"provider\\": \\"userPools\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 2`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"authors\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 3`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"private\\",
  \\"provider\\": \\"userPools\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 4`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"owner\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 5`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"owner\\",
  \\"provider\\": \\"userPools\\",
  \\"ownerFieldName\\": \\"owner\\",
  \\"ownerFieldType\\": \\"string\\",
  \\"identityClaim\\": \\"user_id\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 6`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"ownerContent\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 7`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"owner\\",
  \\"provider\\": \\"userPools\\",
  \\"ownerFieldName\\": \\"authors\\",
  \\"ownerFieldType\\": \\"string[]\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 8`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"ownersContent\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 9`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"private\\",
  \\"provider\\": \\"userPools\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 10`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"privateContent\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 11`] = `
"## [Start] Authorization rules. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $authRules = [] )
#if( $ctx.stash.adminRoles && $ctx.stash.adminRoles.size() > 0 )
  $util.qr($authRules.add({
  \\"provider\\": \\"iam\\",
  \\"type\\": \\"admin\\",
  \\"strict\\": false,
  \\"roles\\": $ctx.stash.adminRoles
}))
#end
$util.qr($authRules.add({
  \\"type\\": \\"public\\",
  \\"provider\\": \\"apiKey\\"
}))
#set( $authResult = $util.authRules.validateUsingSource($authRules, $ctx.source) )
#if( !$authResult )
  $util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization rules. **"
`;

exports[`Verify RDS Model level Auth rules on queries: should successfully transform different field auth rules 12`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source[\\"publicContent\\"])
#end
## [End] Checking for allowed operations which can return this field. **"
`;
